File: src/C#/BehavioralPatterns/State/VendingMachine/PaymentState.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace VendingMachine
{
    public class PaymentState : State
    {
        private decimal funds = 0;

        public PaymentState(VendingMachine vendingMachine) : base(vendingMachine)
        {
            Console.WriteLine($"PAYMENT - You can insert coins.");
        }

        public override void InsertMoney(decimal amount)
        {
            funds += amount;

            var selectedProduct = vendingMachine.Products.FirstOrDefault(p => p.Code == vendingMachine.SelectedProductCode);

            if (funds < selectedProduct.Price)
            {
                Console.WriteLine($"Remaining: {selectedProduct.Price - funds}");
            }
            else
            {
                Console.WriteLine($"Proper amount received");

                var change = funds - selectedProduct.Price;

                if (change > 0)
                {
                    Console.WriteLine($"Dispensing {change} amount");
                }

                vendingMachine.SetState(new DispenseProductState(vendingMachine));
                vendingMachine.DispenseProduct();
            }
        }

        public override void SelectProduct(string productCode) =>
            Console.WriteLine("The product is already selected. Please complete or cancel the current payment");

        public override void DispenseProduct() => Console.WriteLine("Cannot dispense product yet. Insufficient funds.");

        public override void Cancel()
        {
            Console.WriteLine("Cancelling order.");

            if (funds > 0)
            {
                Console.WriteLine($"Returning the amount of {funds}");
            }

            vendingMachine.SelectedProductCode = null;
            vendingMachine.SetState(new IdleState(vendingMachine));
        }

        public override void Refill(List<Product> products) =>
            Console.WriteLine("Cannot refill during payment operation. Please cancel or complete the payment before refill.");
    }
}

File: src/JavaScript/BehavioralPatterns/State/VendingMachine/paymentState.js

const State = require("./state");
const DispenseProductState = require("./dispenseProductState");
const IdleState = require("./idleState");

class PaymentState extends State {
    constructor(vendingMachine) {
        super(vendingMachine);
        this.funds = 0;
        console.log("PAYMENT - You can insert coins.");
    }

    insertMoney(amount) {
        this.funds += amount;

        const selectedProduct = this.vendingMachine.products.find(p => p.code === this.vendingMachine.selectedProductCode);

        if (!selectedProduct) {
            console.log("No product selected.");
            return;
        }

        if (this.funds < selectedProduct.price) {
            console.log(`Remaining: ${selectedProduct.price - this.funds}`);
        } else {
            console.log("Proper amount received");

            const change = this.funds - selectedProduct.price;

            if (change > 0) {
                console.log(`Dispensing ${change} amount`);
            }

            this.vendingMachine.setState(new DispenseProductState(this.vendingMachine));
            this.vendingMachine.dispenseProduct();
        }
    }

    selectProduct(productCode) {
        console.log("The product is already selected. Please complete or cancel the current payment.");
    }

    dispenseProduct() {
        console.log("Cannot dispense product yet. Insufficient funds.");
    }

    cancel() {
        console.log("Cancelling order.");

        if (this.funds > 0) {
            console.log(`Returning the amount of ${this.funds}`);
        }

        this.vendingMachine.selectedProductCode = null;
        this.vendingMachine.setState(new IdleState(this.vendingMachine));
    }

    refill(products) {
        console.log("Cannot refill during payment operation. Please cancel or complete the payment before refill.");
    }
}

module.exports = PaymentState;